File: app/src/main/java/com/zwstudio/logicpuzzlesandroid/puzzles/skyscrapers/domain/SkyscrapersGame.java
package com.zwstudio.logicpuzzlesandroid.puzzles.skyscrapers.domain;

import com.zwstudio.logicpuzzlesandroid.common.domain.CellsGame;
import com.zwstudio.logicpuzzlesandroid.common.domain.GameInterface;
import com.zwstudio.logicpuzzlesandroid.common.domain.Position;
import com.zwstudio.logicpuzzlesandroid.home.domain.HintState;

import java.util.List;

import fj.F2;

/**
 * Created by zwvista on 2016/09/29.
 */

public class SkyscrapersGame extends CellsGame<SkyscrapersGame, SkyscrapersGameMove, SkyscrapersGameState> {
    public static Position offset[] = {
            new Position(-1, 0),
            new Position(0, 1),
            new Position(1, 0),
            new Position(0, -1),
    };

    public boolean isValid(int row, int col) {
        return row >= 1 && col >= 1 && row < size.row - 1 && col < size.col - 1;
    }

    public int[] objArray;
    public int intMax = 1;
    public int get(int row, int col) {
        return objArray[row * cols() + col];
    }
    public int get(Position p) {
        return get(p.row, p.col);
    }
    public void set(int row, int col, int obj) {
        objArray[row * cols() + col] = obj;
    }
    public void set(Position p, int obj) {
        set(p.row, p.col, obj);
    }

    public SkyscrapersGame(List<String> layout, GameInterface<SkyscrapersGame, SkyscrapersGameMove, SkyscrapersGameState> gi) {
        super(gi);
        size = new Position(layout.size(), layout.get(0).length());
        objArray = new int[rows() * cols()];

        for (int r = 0; r < rows(); r++) {
            String str = layout.get(r);
            for (int c = 0; c < cols(); c++) {
                char ch = str.charAt(c);
                int n = ch == ' ' ? 0 : ch - '0';
                set(r, c, n);
                if (intMax < n) intMax = n;
            }
        }

        SkyscrapersGameState state = new SkyscrapersGameState(this);
        states.add(state);
        levelInitilized(state);
    }

    private boolean changeObject(SkyscrapersGameMove move, F2<SkyscrapersGameState, SkyscrapersGameMove, Boolean> f) {
        if (canRedo()) {
            states.subList(stateIndex + 1, states.size()).clear();
            moves.subList(stateIndex, states.size()).clear();
        }
        SkyscrapersGameState state = cloner.deepClone(state());
        boolean changed = f.f(state, move);
        if (changed) {
            states.add(state);
            stateIndex++;
            moves.add(move);
            moveAdded(move);
            levelUpdated(states.get(stateIndex - 1), state);
        }
        return changed;
   }

    public boolean switchObject(SkyscrapersGameMove move) {
        return changeObject(move, (state, move2) -> state.switchObject(move2));
    }

    public boolean setObject(SkyscrapersGameMove move) {
        return changeObject(move, (state, move2) -> state.setObject(move2));
    }

    public int getObject(Position p) {
        return state().get(p);
    }

    public int getObject(int row, int col) {
        return state().get(row, col);
    }

    public HintState getState(int row, int col) {
        return state().getState(row, col);
    }
}

File: app/src/main/java/com/zwstudio/logicpuzzlesandroid/puzzles/skyscrapers/domain/SkyscrapersGame.kt
package com.zwstudio.logicpuzzlesandroid.puzzles.skyscrapers.domain

import com.zwstudio.logicpuzzlesandroid.common.data.GameDocumentInterface
import com.zwstudio.logicpuzzlesandroid.common.domain.CellsGame
import com.zwstudio.logicpuzzlesandroid.common.domain.GameInterface
import com.zwstudio.logicpuzzlesandroid.common.domain.Position

class SkyscrapersGame(layout: List<String>, gi: GameInterface<SkyscrapersGame, SkyscrapersGameMove, SkyscrapersGameState>, gdi: GameDocumentInterface) : CellsGame<SkyscrapersGame, SkyscrapersGameMove, SkyscrapersGameState>(gi, gdi) {
    companion object {
        var offset = arrayOf(
            Position(-1, 0),
            Position(0, 1),
            Position(1, 0),
            Position(0, -1)
        )
    }

    var objArray: IntArray
    fun intMax() = rows - 2

    operator fun get(row: Int, col: Int) = objArray[row * cols + col]
    operator fun get(p: Position) = this[p.row, p.col]
    operator fun set(row: Int, col: Int, obj: Int) {objArray[row * cols + col] = obj}
    operator fun set(p: Position, obj: Int) {this[p.row, p.col] = obj}
    override fun isValid(row: Int, col: Int) = row >= 1 && col >= 1 && row < size.row - 1 && col < size.col - 1

    init {
        size = Position(layout.size, layout[0].length)
        objArray = IntArray(rows * cols)
        for (r in 0 until rows) {
            val str = layout[r]
            for (c in 0 until cols) {
                val ch = str[c]
                val n = if (ch == ' ') 0 else ch - '0'
                this[r, c] = n
            }
        }
        val state = SkyscrapersGameState(this)
        states.add(state)
        levelInitilized(state)
    }

    private fun changeObject(move: SkyscrapersGameMove, f: (SkyscrapersGameState, SkyscrapersGameMove) -> Boolean): Boolean {
        if (canRedo) {
            states.subList(stateIndex + 1, states.size).clear()
            moves.subList(stateIndex, states.size).clear()
        }
        val state = cloner.deepClone(currentState)
        val changed = f(state, move)
        if (changed) {
            states.add(state)
            stateIndex++
            moves.add(move)
            moveAdded(move)
            levelUpdated(states[stateIndex - 1], state)
        }
        return changed
    }

    fun switchObject(move: SkyscrapersGameMove) = changeObject(move, SkyscrapersGameState::switchObject)
    fun setObject(move: SkyscrapersGameMove) = changeObject(move, SkyscrapersGameState::setObject)

    fun getObject(p: Position) = currentState[p]
    fun getObject(row: Int, col: Int) = currentState[row, col]
    fun getState(row: Int, col: Int) = currentState.getState(row, col)
}
